pkg/ccl/changefeedccl/sink_webhook.go

```diff
@@ -220,15 +267,88 @@ func (s *webhookSink) sendMessage(ctx context.Context, reqBody []byte) error {
 	return nil
 }
 
+// workerIndex assigns rows each to a worker goroutine based on the hash of its
+// primary key. This is to ensure that each message with the same key gets
+// deterministically assigned to the same worker. Since we have a channel per
+// worker, we can ensure per-worker ordering and therefore guarantee per-key
+// ordering.
+func (s *webhookSink) workerIndex(key []byte) uint32 {
+	return crc32.ChecksumIEEE(key) % uint32(s.parallelism)
+}
+
+// TODO (ryan min): add memory monitoring for inflight messages
+// inflightTracker wraps logic for counting number of inflight messages to
+// track when flushing sink, with error handling functionality. Implemented
+// as a wrapper for WaitGroup under a lock to block additional messages while
+// flushing.
+type inflightTracker struct {
+	// two mutexes are needed here since different goroutines will be accessing
+	// both the waitgroup and the flushErr at the same time. (i.e. when waiting,
+	// before Done() is called, the error is set, causes deadlock if the same
+	// lock is used)
+	wgMu struct {
+		syncutil.Mutex
+		inflightGroup *sync.WaitGroup
+	}
+	errMu struct {
+		syncutil.Mutex
+		flushErr error
+	}
+}
+
+func makeInflightTracker() *inflightTracker {
+	inflight := &inflightTracker{}
+	inflight.wgMu.inflightGroup = new(sync.WaitGroup)
+	return inflight
+}
+
+// maybeSetError sets flushErr to be err if it has not already been set.
+func (i *inflightTracker) maybeSetError(err error) {
+	i.errMu.Lock()
+	defer i.errMu.Unlock()
+	if err == nil || i.errMu.flushErr != nil {
+		return
+	}
+	i.errMu.flushErr = err
+}
+
+// Add enqueues one inflight message to be flushed.
+func (i *inflightTracker) Add() {
+	i.wgMu.Lock()
+	defer i.wgMu.Unlock()
+	i.wgMu.inflightGroup.Add(1)
+}
+
+// Done tells the inflight tracker one message has been delivered.
+func (i *inflightTracker) Done() {
+	i.wgMu.inflightGroup.Done()
+}
+
+// Wait waits for all inflight messages to be delivered (inflight = 0) and
+// returns a possible error
+func (i *inflightTracker) Wait() error {
+	i.wgMu.Lock()
+	defer i.wgMu.Unlock()
+	defer i.errMu.Unlock()
+	i.wgMu.inflightGroup.Wait()
+	// lock here to avoid deadlock with wgMu
```

Comment:
The deadlock here is that the requests we are Wait'ing on may need to set the error.  If that is correct, let's spell that out explicitly.